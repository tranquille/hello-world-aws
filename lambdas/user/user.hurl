POST https://4rpurhovj2.execute-api.eu-central-1.amazonaws.com/stage/user
{ 
 "token": "eyJraWQiOiJFMlBRUVNJNE9lQXlEWmFkWUQrVkZzaHFMOVwveDJPa0FLNlM3MXFLRmxzbz0iLCJhbGciOiJSUzI1NiJ9.eyJzdWIiOiJhMGVmZDE2Zi03OTRiLTQ5MzQtOTIzYi0xODBhYTQzYzFmMGIiLCJpc3MiOiJodHRwczpcL1wvY29nbml0by1pZHAuZXUtY2VudHJhbC0xLmFtYXpvbmF3cy5jb21cL2V1LWNlbnRyYWwtMV95WnNDck1CRGUiLCJjbGllbnRfaWQiOiIzcDhiYTZoYzN1OWhuZmIxaWczYjNrMGNqdCIsIm9yaWdpbl9qdGkiOiI2ZGQ1NjlkNi01ZDZiLTRjOGQtYTQxNi1hMjg3Nzg1MzE5MDQiLCJldmVudF9pZCI6IjVjZjRkNzE0LWExYTQtNDU3MC1iM2MxLWQyYjYxNmM2ZTU0YiIsInRva2VuX3VzZSI6ImFjY2VzcyIsInNjb3BlIjoiYXdzLmNvZ25pdG8uc2lnbmluLnVzZXIuYWRtaW4iLCJhdXRoX3RpbWUiOjE2ODEyMTI2OTcsImV4cCI6MTY4MTIxNjI5NywiaWF0IjoxNjgxMjEyNjk3LCJqdGkiOiIzOTRiODAyYi0xMTE1LTRmMmMtOTNlNy02MjI1MDkxNDk0MjUiLCJ1c2VybmFtZSI6ImEwZWZkMTZmLTc5NGItNDkzNC05MjNiLTE4MGFhNDNjMWYwYiJ9.qmuBG_BpH3EIypf3C-wF39G02b7DGWhLgcWzof4mGgY4kdqamNk7Lm5WSB0TiRpQ1UYX3PsBhI-8kngeWhj8mkUl62O124PnakB3l4AcCJLm32V3qDirYmVjqtz-hL41vMcyy6MGZrJrPtbFt-sh2n7MwsRlT12gJE7uR2hXoNvWO2qhYYB7YDoG7FLNEv2rLTh99oywnR7K6Dyc5O4XNYKnJYElQZj2t1vhAJKcS9h5z87lnOxTMOB-4nfOhako4xwP6JPEXdM9S_e_8dsZqw7Wty7o9RPvCOJn0l03IutPYOUOGNKfLnEnqkhggb2Cytf0EQxcV1aWEIE5DMbOag"
}