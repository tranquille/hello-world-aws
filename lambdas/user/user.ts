import { DynamoDBClient, GetItemCommand } from "@aws-sdk/client-dynamodb";
import { unmarshall } from "@aws-sdk/util-dynamodb";
import {
  CognitoIdentityProviderClient,
  GetUserCommand,
} from "@aws-sdk/client-cognito-identity-provider";
import { APIGatewayProxyEvent, APIGatewayProxyResult } from "aws-lambda";

const documentClient = new DynamoDBClient({});
const identityProviderClient = new CognitoIdentityProviderClient({});

const TABLE_NAME = process.env.USER_DB_TABLE_NAME;

enum HttpStatusCode {
  OK = 200,
  CREATED = 201,
  NO_CONTENT = 204,
  NOT_FOUND = 404,
  BAD_REQUEST = 400,
  INTERNAL_SERVER_ERROR = 500,
}

const response = (statusCode: HttpStatusCode, body?: unknown) => {
  return {
    statusCode,
    headers: {
      "Content-Type": "application/json",
    },
    isBase64Encoded: false,
    body: JSON.stringify(body),
  };
};

const isNotEmpty = (s: string | undefined | null): boolean =>
  s !== undefined && s !== null && typeof s === "string" && s.length > 0;

const isEmpty = (s: string): boolean => !isNotEmpty(s);

export const handler = async (
  event: APIGatewayProxyEvent
): Promise<APIGatewayProxyResult> => {
  try {
    const { body, isBase64Encoded } = event;
    if (body === undefined || body === null) {
      return response(HttpStatusCode.BAD_REQUEST, "body is not defined");
    }

    let parsedBody;
    if (isBase64Encoded === true) {
      const buff = Buffer.from(body, "base64");
      const eventBodyStr = buff.toString("utf8");
      parsedBody = JSON.parse(eventBodyStr);
    } else {
      parsedBody = JSON.parse(body);
    }
    const { token } = parsedBody;

    if (isEmpty(token)) {
      return response(
        HttpStatusCode.BAD_REQUEST,
        "provided payload is not valid"
      );
    }

    const getUserCommand = new GetUserCommand({
      AccessToken: token,
    });
    const user = await identityProviderClient.send(getUserCommand);
    console.log("Logged in successfully");
    console.log(user);

    let item;
    const email = user.UserAttributes?.find(
      (attr) => attr.Name === "email"
    )?.Value;

    if (email) {
      const command = new GetItemCommand({
        TableName: TABLE_NAME,
        Key: {
          email: {
            S: email,
          },
        },
      });
      const { Item } = await documentClient.send(command);
      item = Item !== undefined ? unmarshall(Item) : undefined;
      console.log(item, Item);
      console.log("User Data retrieved successfully");
    }

    return response(HttpStatusCode.OK, item);
    // rome-ignore lint/suspicious/noExplicitAny: Catch clause accepts only any or unknown
  } catch (e: any) {
    const statusCode =
      e?.$metadata?.httpStatusCode || HttpStatusCode.INTERNAL_SERVER_ERROR;

    console.error("Lambda Function failed with ", e);
    return response(statusCode, e);
  }
};
