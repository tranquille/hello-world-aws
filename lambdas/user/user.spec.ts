vi.mock("@aws-sdk/util-dynamodb");

const identityProviderSend = vi.fn();
vi.mock("@aws-sdk/client-cognito-identity-provider", async () => {
  // rome-ignore lint/suspicious/noExplicitAny: <explanation>
  const actual: any = await vi.importActual(
    "@aws-sdk/client-cognito-identity-provider",
  );

  const CognitoIdentityProviderClient = vi.fn(() => ({
    send: identityProviderSend,
  }));

  return { ...actual, CognitoIdentityProviderClient };
});

const dynamoDbSend = vi.fn();
vi.mock("@aws-sdk/client-dynamodb", async () => {
  // rome-ignore lint/suspicious/noExplicitAny: <explanation>
  const actual: any = await vi.importActual("@aws-sdk/client-dynamodb");

  const DynamoDBClient = vi.fn(() => ({
    send: dynamoDbSend,
  }));

  return { ...actual, DynamoDBClient };
});

import { APIGatewayProxyEvent } from "aws-lambda";
import { handler } from "./user";

const getTestEvent = (_payload: unknown, isBase64Encoded = false) => {
  let payload = JSON.stringify(_payload);
  if (isBase64Encoded) {
    payload = Buffer.from(JSON.stringify(_payload)).toString("base64");
  }

  return {
    body: payload,
    headers: {},
    multiValueHeaders: {},
    httpMethod: "POST",
    isBase64Encoded,
    path: "",
    pathParameters: null,
    queryStringParameters: null,
    multiValueQueryStringParameters: null,
    stageVariables: null,
    // rome-ignore lint/suspicious/noExplicitAny: It's a test
    requestContext: {} as any,
    resource: "",
  };
};

describe("Login Lambda", () => {
  beforeEach(() => {
    vi.resetAllMocks();
  });

  it("retrieves the user data if a AuthenticationToken is provided", async () => {
    identityProviderSend.mockReturnValue({
      UserAttributes: [
        {
          Name: "email",
          Value: "asdf@asdf.de",
        },
      ],
    });
    dynamoDbSend.mockReturnValue({
      Item: {
        SS: ["Volleyball", "Basketball"],
      },
    });
    const testEvent: APIGatewayProxyEvent = getTestEvent({
      token: "asdsadsadasdsadsad",
    });

    const result = await handler(testEvent);

    expect(identityProviderSend).toHaveBeenCalled();

    expect(dynamoDbSend).toHaveBeenCalled();

    expect(result).toMatchSnapshot();
  });
});
