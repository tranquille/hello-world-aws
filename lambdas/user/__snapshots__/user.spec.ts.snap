// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Login Lambda > retrieves the user data if a AuthenticationToken is provided 1`] = `
{
  "body": undefined,
  "headers": {
    "Content-Type": "application/json",
  },
  "isBase64Encoded": false,
  "statusCode": 200,
}
`;
